# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [28]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [29]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [30]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [31]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [32]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [33]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [34]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [35]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [36]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [37]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [38]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [39]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [40]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [41]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [42]:
data.duplicated().sum()

71

In [43]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [44]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [45]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [46]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [47]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [48]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [49]:
data.groupby('children')['debt'].mean()

children
0    0.075438
1    0.092346
2    0.094542
3    0.081818
4    0.097561
5    0.000000
Name: debt, dtype: float64

**Вывод:** 

Как можем заметить, риск не выплатить долг растёт вместе с увеличением количества детей. Вместе с тем люди, имеющие пятерых детей выплатили 100% кредитов, а люди, имеющие троих выплачивают его чаще тех, у кого один ребёнок: впрочем, в выборке присутствует гораздо меньше значений для семей от троих детей, нежели для бездетных или тех, у кого 1-2 ребёнка, что делает это сравнение неоднозначным. 
Можно сделать вывод, что люди без детей имеют меньшие риски просрочить оплату нежели те, у кого есть один ребёнок и больше. 

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [50]:
# создадим новый датафрейм для отражения отношения выплаты долга к семейному положению

debt_to_status = pd.DataFrame() 
# подсчитаем общее число займов, разделённое на группы по семейному положению

debt_to_status['debt_total'] = data.groupby('family_status')['debt'].count()

# выведем значение для количества неоплаченных кредитов, разделённое на группы по семейному положению

debt_to_status['debt_unpaid'] = data.groupby('family_status')['debt'].sum() 

# подсчитаем отношение наличия долга к общему числу займов

debt_to_status['dependency_score_status'] = data.groupby('family_status')['debt'].sum() / \
data.groupby('family_status')['debt'].count() * 100 
debt_to_status.sort_values('dependency_score_status', ascending = False)

debt_total  debt_unpaid  dependency_score_status
family_status                                                          
Не женат / не замужем        2796          273                 9.763948
гражданский брак             4134          385                 9.313014
женат / замужем             12261          927                 7.560558
в разводе                    1189           84                 7.064760
вдовец / вдова                951           63                 6.624606

**Вывод:** 

На основании предоставленных данных можно сделать вывод, что зависимость выплаты кредита в срок от семейного положения существует:
* люди, прежде не состоявшие в официальном браке, имеют просрочку по задолженности в гораздо большем количестве случаев 
* состоящие, либо состоявшие в браке ранее, но сменившие семейное положение в силу обстоятельств (развод/смерть супруга) имеют тенденцию к снижению риска задержать платёж
* при этом наименьший риск задолженности наблюдается в группе вдовцов

Т.к. количество наблюдений в группах выборки распределено с незначительной разницей (не считая группы женатых/замужних людей), можно сделать выводы о релевантности семейного положения в отношении наличия долга по кредиту. 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [51]:
# воспользуемся сформированными ранее категориями дохода

data.groupby('total_income_category')['total_income_category'].count()

total_income_category
A       25
B     5014
C    15921
D      349
E       22
Name: total_income_category, dtype: int64

Как можно заметить, в выборке снова наблюдается большой разброс в репрезентативности групп, отчего можно предположить, что сформировать корректные выводы по ним будет проблематично. Проверим эту гипотезу. 

In [52]:
# воспользуемся другими методами формирования отчёта 
# формируем таблицу с помощью группировки и объединения

data_debt_sum = data.groupby(['total_income_category']).agg({'debt':'sum'})
data_debt_count = data.groupby(['total_income_category']).agg({'debt':'count'})
debt_to_income = data_debt_sum.merge(data_debt_count, on='total_income_category', how='left') 
debt_to_income['процент невыплаты долга'] = debt_to_income['debt_x'] / debt_to_income['debt_y'] * 100
debt_to_income.rename(columns={'debt_x': 'число задолженостей', 'debt_y' : 'число займов'}).sort_values('процент невыплаты долга', ascending=False)

число задолженостей  число займов  \
total_income_category                                      
E                                        2            22   
C                                     1353         15921   
A                                        2            25   
B                                      354          5014   
D                                       21           349   

                       процент невыплаты долга  
total_income_category                           
E                                     9.090909  
C                                     8.498210  
A                                     8.000000  
B                                     7.060231  
D                                     6.017192

**Вывод:** 

Несмотря на малую репрезентативность некоторых категорий, можно сделать следующие выводы: группы с самым высоким доходом - **A**, и самым низким - **E**, несут за собой примерно одинаковое количество заёмщиков и одинаковое количество должников, что делает зависимость невыплаты долга от уровня дохода потенциально незначимой. Вместе с тем:
* группа **C**, представляющая собой людей со средним доходом и являющаяся самой крупной, находится на втором месте по количеству должников.
* группы **B** и **D**, отвечающие за клиентов с уровнем дохода выше и ниже среднего соответственно, занимают нижние места таблицы и возвращают долг в срок чаще других.

Так как самыми надёжными заёмщиками являются люди с уровнем дохода ниже среднего, сложно сделать вывод о том, что уровень дохода напрямую влияет на выплату в срок: уместно рассматривать этот критерий вместе с другими - например, возрастом или наличием детей.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [53]:
# используем сводную таблицу для формирования отчёта

data_pivot_1 = data.pivot_table(index=['purpose_category'], values='debt', aggfunc='sum')
data_pivot_2 = data.pivot_table(index=['purpose_category'], values='debt', aggfunc='count')
debt_to_purpose = data_pivot_1.merge(data_pivot_2, on='purpose_category', how='left')
debt_to_purpose['процент невыплаты долга'] = debt_to_purpose['debt_x'] / debt_to_purpose['debt_y'] * 100
debt_to_purpose.rename(columns={'debt_x': 'число задолженостей', 'debt_y' : 'число займов'}).sort_values('процент невыплаты долга', ascending=False)

число задолженостей  число займов  \
purpose_category                                              
операции с автомобилем                    400          4279   
получение образования                     369          3988   
проведение свадьбы                        183          2313   
операции с недвижимостью                  780         10751   

                          процент невыплаты долга  
purpose_category                                   
операции с автомобилем                   9.347978  
получение образования                    9.252758  
проведение свадьбы                       7.911803  
операции с недвижимостью                 7.255139

**Вывод:** 

Категория людей, берущих кредит на операции с недвижимостью, является самой надёжной в отношении погашения долга в срок. Вместе с тем люди, обращающиеся за займом в целях связанных с образованием или транспортным средством, демонстрируют увеличенную тенденцию к просрочке выплат.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* если судить по столбцам с пропущенными значениями (стаж и общий уровень дохода), можно предположить, что для выдачи некоторых кредитов указание этих данных было необязательным.

Так же никогда нельзя исключать техническую сторону вопроса и то, что при заполнении данных сотрудником кредитного отдела могли быть допущены ошибки, что привело к ошибке чтения программой. Однако, учитывая условия выдачи займов в некоторых кредитных организациях, а так же то, что количество пропусков в этих столбцах было одинаковым, склоняюсь к первому варианту.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* в выборке с количественными переменными может наблюдаться сильное отклонение от подавляющего большинства значений, или выбросы - как в меньшую, так и в большую стороны, - что может привести к неверным выводам при использовании среднего арифметического.

Например, если взять выборку натуральных чисел от 1 до 100, и предположить, что 95% этих чисел находятся в диапазоне от 40 до 60, но при этом наблюдается определённое количество значений в виде единиц и сотен, взятие медианы из этой выборки даст более полное представление о ней, нежели выведение среднего.

### Шаг 4: общий вывод.

Цель данного исследования - провести первичную оценку платёжеспособности заемщиков. Проведённый анализ факторов, влияющих на факт погашения долга в срок, поможет в формировании новой модели кредитного скоринга, способной сопоставить потенциальных клиентов с определённым портретом заёмщика, полученным по его итогам.

**Выделим следующие гипотезы:**
1. Количество детей влияет на погашение кредита в срок;
2. Семейное положение влияет на погашение кредита в срок;
3. Уровень дохода влияет на погашение кредита в срок;  
4. Цель кредита влияет на его погашение в срок.

В результате анализа удалось выявить следующее: 

**Первая гипотеза частично подтверждена.**

Среди выборок, количество данных в которых позволяет сделать более обоснованные выводы, выделим группы без детей, с 1 ребёнком или двумя детьми. Для них:

* доля просроченных выплат по группе с 2 детьми -  **9,4%**;
* доля просроченных выплат по группе с 1 ребёнком -  **9,2%**;
* доля просроченных выплат по группе без детей составила **7,5%**;

Можно заметить, что наблюдается большой скачок (**1,7%**) в просрочке платежей между семьями без детей и семьями, в которых появился ребёнок. Для последующего ребёнка увеличение составляет лишь 0,2%. К сожалению, составить более детальный вывод по вводным данным невозможно - требуется исследование с большим количеством семей, представляющих группы из 3, 4 и 5 детей. 
По предоставленным данным можно сделать лишь вывод о том, что заёмщики без детей выплачивают кредит в срок в большем количестве случаев. 

Рекомендация: для возможности полноценного использования этой гипотезы, требуется исследование на большем количестве данных.

**Вторая гипотеза подтверждена.**

* для людей, никогда не состоявших в официальных отношениях доля задолженности - самая высокая: не женаты/не замужем - **9,7%**, состоят в гражданском браке - **9,3%**;
* люди, состоящие в браке имеют более низкий процент просроченных платежей - **7,5**;
* в то же время люди, состоявшие в браке ранее, - развод (**7%**) и вдовец/вдова (**6.6%**), - обладают наименьшей долей задолженности.

Рекомендация: при оценке платежеспособности клиентов в условии значимости любых значений, разница между группами людей, не состоявших в официальных отношениях и состоявших когда-либо (**2%**), может использоваться в построении модели скоринга. 

**Третья гипотеза не подтверждена.**

Отсутствие достаточных данных в группах людей с очень низким - **E** (**<30000**) и очень высоким доходом - **A** (**>1000000**) не позволяет использовать их для проведения достоверного анализа. Вместе с тем:

* люди со средним доходом (категория **C**, **50001–200000**) имеют самый высокий (**8.4**) процент задолженности среди оставшихся категорий;
* группа **B** (**200001-1000000**) находится на втором месте, с долей задолженности в **7%**, а группа **D**, представленная доходом ниже среднего (**30001-50000**) - **6%**.

Рекомендация: не использовать фактор дохода как самостоятельную единицу в исследовании, объединить его с другими факторами - такими как возраст, количество детей и уровень образования.

**Гипотеза 4 подтверждена.**
      
При анализе четвёртого фактора наблюдается следующее:

* кредиты, взятые на операции с автомобилем, были просрочены в **9.3%** случаев;
* кредиты, взятые на получение образования, были просрочены в **9.2%** случаев;
* кредиты, взятые на проведение свадьбы, были просрочены в **7.9%** случаев;
* кредиты, взятые на операции с недвижимостью, были просрочены в **7.2%** случаев.

Рекомендация: беря во внимание тот же фактор, что и при проверке второй гипотезы - любая разница в значениях считается значимой, можно сделать вывод, что люди, берущие кредит на операции с недвижимостью являются более надёжными заёмщиками нежели те, что берут его для покупки автомобиля или в целях трат на обучение, и использовать эти данные в построении модели скоринга.

**Общие выводы:**
1. Для проверки первой гипотезы требуется более объёмная выборка.
2. Зависимости в данных из второй и четвёртой гипотез могут быть использованы при формировании потрета заёмщика, а так же моделировании системы кредитного скоринга.
3. Не стоит использовать зависимости из третьей гипотезы как самостоятельную категорию оценки платежеспособности - правильнее объеденить её с другими категориями. 
